### PR TITLE
Fixes incomplete promise when a child verticle could not be deployed …

### PR DESCRIPTION
**Pull request by zenios**
…because its parent has been undeployed

Signed-off-by: zenios <dimitris.zenios@gmail.com>

Not sure if its better to wait for underemployment to complete and then fail the promise.

### CODE REVIEW

**Comment by vietj** [comment]
can you write a test for this ?

**Comment by zenios** [comment]
I will try

As for the promise do you think its better to wait for the undeployment to
complete or its ok like this?

**Comment by vietj** [comment]
let's write a test for this first :-)

**Comment by vietj** [comment]
that will help having a decision

**Comment by zenios** [comment]
Tests added

File: src/main/java/io/vertx/core/impl/DeploymentManager.java [hunk 1]
```diff
@@ -463,7 +463,8 @@ private ClassLoader getCurrentClassLoader() {
                   deployment.child = true;
                 } else {
                   // Orphan
-                  deployment.undeploy(null);
+                  deployment.undeploy(event -> {});
```
Comment (vietj) [inline src/main/java/io/vertx/core/impl/DeploymentManager.java]: I think we should fail the promise in the undeploy callback

**Comment by zenios** [comment]
Ok let me change it

**Comment by zenios** [comment]
Done

File: src/test/java/io/vertx/core/DeploymentTest.java [hunk 2]
```diff
@@ -1234,6 +1234,28 @@ public void testUndeployAll() throws Exception {
...
+        vertx.deployVerticle(ChildVerticle.class.getName())
+          .<Void>mapEmpty()
+          .setHandler(stopPromise);
+      }
+    }
+
+    CountDownLatch latch = new CountDownLatch(1);
+    vertx.deployVerticle(new ParentVerticle(), ar1 -> {
+      assertTrue(ar1.succeeded());
+      vertx.undeploy(ar1.result(), ar2 -> {
```
Comment (vietj) [inline src/test/java/io/vertx/core/DeploymentTest.java]: you can use the helper method onSuccess / onFailure to make the test shorter and easier to read

Comment (zenios) [reply src/test/java/io/vertx/core/DeploymentTest.java]: Done

File: src/test/java/io/vertx/core/DeploymentTest.java [hunk 3]
```diff
@@ -1234,6 +1234,23 @@ public void testUndeployAll() throws Exception {
...
+        vertx.deployVerticle(ChildVerticle.class.getName())
+          .<Void>mapEmpty()
+          .setHandler(stopPromise);
+      }
+    }
+
+    CountDownLatch latch = new CountDownLatch(1);
+    vertx.deployVerticle(new ParentVerticle(), onSuccess(id ->
+      vertx.undeploy(id, onFailure(u -> latch.countDown()))));
+    awaitLatch(latch);
```
Comment (vietj) [inline src/test/java/io/vertx/core/DeploymentTest.java]: instead use `await()` and `testComplete()` to signal test completion

**Comment by vietj** [comment]
another comment and I think we are good

File: src/test/java/io/vertx/core/DeploymentTest.java (hunk 3 above)
Comment (zenios) [reply src/test/java/io/vertx/core/DeploymentTest.java]: Done.Do you want me to squash to single commit?

Comment (vietj) [reply src/test/java/io/vertx/core/DeploymentTest.java]: thanks, no I'll do it on merge

**Pull request merged**